### ISSUE TITLE
Failed to build GRPC with tf-nightly

### ISSUE THREAD

**Issue by jrabary** [google-research/seed_rl#14]
Dear all,

I'm trying to build de GRPC package with tf-nightly but got the following error:

```
ERROR: /custom-op/grpc/BUILD.bazel:4:1: C++ compilation of rule '//grpc:ops/grpc.so' failed (Exit 1)
In file included from bazel-out/k8-opt/bin/grpc/service.grpc.pb.h:21:0,
                 from grpc/ops/grpc.cc:23:
bazel-out/k8-opt/bin/grpc/service.pb.h:17:2: error: #error This file was generated by an older version of protoc which is
 #error This file was generated by an older version of protoc which is
  ^~~~~
bazel-out/k8-opt/bin/grpc/service.pb.h:18:2: error: #error incompatible with your Protocol Buffer headers. Please
 #error incompatible with your Protocol Buffer headers. Please
  ^~~~~
bazel-out/k8-opt/bin/grpc/service.pb.h:19:2: error: #error regenerate this file with a newer version of protoc.
 #error regenerate this file with a newer version of protoc.
  ^~~~~
grpc/ops/grpc.cc: In member function 'virtual void tensorflow::{anonymous}::CreateGrpcClientOp::Compute(tensorflow::OpKernelContext*)':
grpc/ops/grpc.cc:1055:23: warning: comparison between signed and unsigned integer expressions [-Wsign-compare]
     for (int i = 0; i < method_output_signatures_list.size(); ++i) {
                     ~~^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
```

Here is the docker file I use:

```
FROM tensorflow/tensorflow:nightly-custom-op-ubuntu16 as grpc_compile

RUN git clone https://github.com/tensorflow/custom-op.git
WORKDIR custom-op

RUN ./configure.sh

RUN echo '\n\
load("@bazel_tools//tools/build_defs/repo:http.bzl", "http_archive")\n\
\n\
http_archive(\n\
    name = "com_github_grpc_grpc",\n\
    urls = [\n\
        "https://github.com/grpc/grpc/archive/ac1c5de1b36da4a1e3d72ca40b0e43f24266121a.tar.gz",\n\
    ],\n\
    strip_prefix = "grpc-ac1c5de1b36da4a1e3d72ca40b0e43f24266121a",\n\
)\n\
\n\
load("@com_github_grpc_grpc//bazel:grpc_deps.bzl", "grpc_deps")\n\
grpc_deps()\n\
load("@com_github_grpc_grpc//bazel:grpc_extra_deps.bzl", "grpc_extra_deps")\n\
grpc_extra_deps()' >> WORKSPACE

ADD grpc/ grpc/
RUN bazel build grpc:ops/grpc.so grpc:service_py_proto --incompatible_remove_legacy_whole_archive=0

ADD . /seed_rl
RUN cp bazel-bin/grpc/ops/grpc.so /seed_rl/grpc/grpc_cc.so
RUN cp bazel-bin/grpc/service_pb2.py /seed_rl/grpc/service_pb2.py
WORKDIR /seed_rl/

RUN PYTHONPATH=/ python grpc/python/ops_test.py
```

**Comment by lespeholt** [comment]
Is there a good reason to compile with nightly?

One suggested would be to update the grpc package (i.e. a newer grpc github hash)

**Comment by jrabary** [comment]
> Is there a good reason to compile with nightly?
I'm trying to use seed rl with tf_agents. My codes needs at least TF2.2 and custom ops docker is only nightly or 2.1

**Comment by jrabary** [comment]
> One suggested would be to update the grpc package (i.e. a newer grpc github hash)

How one can find this ?

**Comment by lespeholt** [comment]
update this rule in your script

RUN echo '\n\
load("@bazel_tools//tools/build_defs/repo:http.bzl", "http_archive")\n\
\n\
http_archive(\n\
    name = "com_github_grpc_grpc",\n\
    urls = [\n\
        "https://github.com/grpc/grpc/archive/ac1c5de1b36da4a1e3d72ca40b0e43f24266121a.tar.gz",\n\
    ],\n\
    strip_prefix = "grpc-ac1c5de1b36da4a1e3d72ca40b0e43f24266121a",\n\
)\n\
\n\
load("@com_github_grpc_grpc//bazel:grpc_deps.bzl", "grpc_deps")\n\
grpc_deps()\n\
load("@com_github_grpc_grpc//bazel:grpc_extra_deps.bzl", "grpc_extra_deps")\n\
grpc_extra_deps()' >> WORKSPACE

**Comment by jrabary** [comment]
Unfortunately it gives the same error.

**Comment by lespeholt** [comment]
hhm, was thinking this could be caused by an old version of bazel but the docker image does seem to have 2.0.0 which also appears to be the recommended for TF

**Comment by lespeholt** [comment]
I have updated SEED to TF 2.2.0 now.

**Comment by jrabary** [comment]
@lespeholt thank you. I'm trying to understand what was the fix but it seems magical to me. Is it because the docker image of custom-ops 2.2.0 was updated ? For example, I'm trying to build it with the nightly version and it still don't work and gives the same error as before. I also tried to build against 2.3.0rc0 with the following docker file

```
FROM tensorflow/tensorflow:2.2.0-custom-op-gpu-ubuntu16 as grpc_compile

RUN pip3 install --upgrade tensorflow-gpu==2.3.0rc0

RUN git clone https://github.com/tensorflow/custom-op.git
WORKDIR custom-op

RUN ./configure.sh

RUN echo '\n\
load("@bazel_tools//tools/build_defs/repo:http.bzl", "http_archive")\n\
\n\
http_archive(\n\
    name = "com_github_grpc_grpc",\n\
    urls = [\n\
        "https://github.com/grpc/grpc/archive/ac1c5de1b36da4a1e3d72ca40b0e43f24266121a.tar.gz",\n\
    ],\n\
    strip_prefix = "grpc-ac1c5de1b36da4a1e3d72ca40b0e43f24266121a",\n\
)\n\
\n\
load("@com_github_grpc_grpc//bazel:grpc_deps.bzl", "grpc_deps")\n\
grpc_deps()\n\
load("@com_github_grpc_grpc//bazel:grpc_extra_deps.bzl", "grpc_extra_deps")\n\
grpc_extra_deps()' >> WORKSPACE

ADD grpc/ grpc/
RUN bazel build grpc:ops/grpc.so grpc:service_py_proto --incompatible_remove_legacy_whole_archive=0

ADD . /seed_rl
RUN cp bazel-bin/grpc/ops/grpc.so /seed_rl/grpc/grpc_cc.so
RUN cp bazel-bin/grpc/service_pb2.py /seed_rl/grpc/service_pb2.py
WORKDIR /seed_rl/

RUN pip3 install tensorflow-gpu==2.3.0rc0

RUN PYTHONPATH=/ python3 grpc/python/ops_test.py
```

For the nightly test I used `FROM tensorflow/tensorflow:nightly-custom-op-gpu-ubuntu16 as grpc_compile`

In the commit that fixes the problem with 2.2 I didn't seed any change related to `protoc` and I can't figure out what was the real problem.

**Comment by lespeholt** [comment]
This change was essential to make 2.2.0 work:
https://github.com/tensorflow/custom-op/commit/79edd2d74b9659775232cf0c4fcd3c1d67c65952

**Comment by jrabary** [comment]
@lespeholt I finally got it worked. The problem is that in order to build the seed_rl/grpc package we have to use the same version of protobuf as used for tensorfow. For `TF-2.3+` the protobuf version is `3.9.1`. Below is the fix to build it with `TF-2.3.0rc0`

```
FROM tensorflow/tensorflow:2.2.0-custom-op-gpu-ubuntu16 as grpc_compile

RUN pip3 install tensorflow-gpu==2.3.0rc0

RUN git clone https://github.com/tensorflow/custom-op.git

WORKDIR custom-op

RUN ./configure.sh

RUN echo '\n\
load("@bazel_tools//tools/build_defs/repo:http.bzl", "http_archive")\n\
\n\
http_archive(\n\
    name = "com_google_protobuf",\n\
    strip_prefix = "protobuf-3.9.1",\n\
    urls = [\n\
      "https://github.com/protocolbuffers/protobuf/archive/v3.9.1.tar.gz"\n\
    ],\n\
)\n\
\n\
http_archive(\n\
    name = "com_github_grpc_grpc",\n\
    urls = [\n\
        "https://github.com/grpc/grpc/archive/ac1c5de1b36da4a1e3d72ca40b0e43f24266121a.tar.gz",\n\
    ],\n\
    strip_prefix = "grpc-ac1c5de1b36da4a1e3d72ca40b0e43f24266121a",\n\
)\n\
\n\
load("@com_github_grpc_grpc//bazel:grpc_deps.bzl", "grpc_deps")\n\
grpc_deps()\n\
load("@com_github_grpc_grpc//bazel:grpc_extra_deps.bzl", "grpc_extra_deps")\n\
grpc_extra_deps()' >> WORKSPACE

ADD grpc/ grpc/
RUN bazel build grpc:ops/grpc.so grpc:service_py_proto --incompatible_remove_legacy_whole_archive=0

ADD . /seed_rl
RUN cp bazel-bin/grpc/ops/grpc.so /seed_rl/grpc/grpc_cc.so
RUN cp bazel-bin/grpc/service_pb2.py /seed_rl/grpc/service_pb2.py
WORKDIR /seed_rl/
RUN pip3 install tensorflow-gpu==2.3.0rc0
RUN PYTHONPATH=/ python3 grpc/python/ops_test.py
```

In the Dockerfile, I change the default target to build protobuf.

**Comment by lespeholt** [comment]
Great :-) I can update it to 2.3 when it's released
